### PR TITLE
#478 total receivable

### PR DESCRIPTION
**Pull request by Satendra-SR**
#478 

### What I did:

- create a method `filterByEndDates` in `Invoice` Modal.
- enhanced my previous mistake;
` $unpaidInvoices = Invoice::where('status', 'unpaid')->orWhere(function ($query) use ($endDate) {
           $query->where('sent_on', '<=', $endDate);
         })->get();`

and replaced with

 `$invoices = Invoice::filterByEndDate($endDate);`.
- previously, it's functionality was not fulfill the requirement need. means it was not performed as expected. this time. I solved it.

### CODE REVIEW

**Comment by Satendra-SR** [comment]
@bohraji 
one thing I want to ask you that if you notice I have used two `for` loop. and it can be optimized.
but the first loop
` foreach (config('constants.currency') as $currency => $currencyMeta) {
              $attr['totalReceivables'][$currency] = 0;
          }`
is running just twice. and if I put it inside the second loop, it will always check the condition.
so what should be the optimum way?

File: app/Http/Controllers/Finance/ReportsController.php [hunk 1]
```diff
@@ -75,14 +75,14 @@ public static function getReportAttributes()
...
-            $query->where('sent_on', '<=', $endDate);
-        })->get();
+        $invoices = Invoice::filterByEndDates($endDate);
         foreach (config('constants.currency') as $currency => $currencyMeta) {
             $attr['totalReceivables'][$currency] = 0;
         }
-        foreach ($unpaidInvoices as $unpaidInvoice) {
-            $attr['totalReceivables'][$unpaidInvoice->currency_sent_amount] += $unpaidInvoice->sent_amount;
+        foreach ($invoices as $invoice) {
+            if ($invoice->status != 'paid') {
```
Comment (bohraji) [inline app/Http/Controllers/Finance/ReportsController.php]: @Satendra-SR In earlier implementation, we were only fetching the unpaid invoices. But in this updated implementation, we are fetching all the invoices before the date and then using only the `not - paid` ones. Please update to filtering only unpaid invoices when fetching it from database.

File: app/Models/Finance/Invoice.php [hunk 2]
```diff
@@ -77,4 +77,13 @@ public function getClientAttribute()
     {
         return optional($this->projectStageBillings()->first()->projectStage->project)->client;
     }
+
+    public static function filterByEndDates($end, $paginated = false)
+    {
+        $invoices = self::with('projectStageBillings.projectStage.project.client')
+            ->whereDate('sent_on', '<=', $end)
```
Comment (bohraji) [inline app/Models/Finance/Invoice.php]: This filter is same as the one above where we are doing filtering using both `start date` and `end date`. Can we have a common query scope for both of them?

File: app/Http/Controllers/Finance/ReportsController.php (hunk 1 above)
Comment (Satendra-SR) [reply app/Http/Controllers/Finance/ReportsController.php]: sure, I was trying to filter it in the controller. but I think that was a bad practice.
thanks

File: app/Models/Finance/Invoice.php (hunk 2 above)
Comment (Satendra-SR) [reply app/Models/Finance/Invoice.php]: sure.

File: app/Models/Finance/Invoice.php [hunk 3]
```diff
@@ -77,4 +77,13 @@ public function getClientAttribute()
     {
         return optional($this->projectStageBillings()->first()->projectStage->project)->client;
     }
+
+    public static function filterByEndDates($end, $paginated = false)
```
[the remark this answers is not in the document]
Comment (Satendra-SR) [reply app/Models/Finance/Invoice.php]: right. it makes sense. I think I forgot that one.

File: app/Models/Finance/Invoice.php [hunk 4]
```diff
@@ -78,12 +78,15 @@ public function getClientAttribute()
...
     }
 
-    public static function filterByEndDates($end, $paginated = false)
+    public static function filterByEndDate($date, $paginated = false)
     {
-        $invoices = self::with('projectStageBillings.projectStage.project.client')
-            ->whereDate('sent_on', '<=', $end)
-            ->orderBy('sent_on', 'desc');
+        $unpaidInvoices = self::where(function ($query) use ($date) {
+            $query->where('sent_on', '<=', $date)
```
Comment (bohraji) [inline app/Models/Finance/Invoice.php]: Let's make this a query scope and reuse it here.

File: app/Models/Finance/Invoice.php [hunk 5]
```diff
@@ -78,12 +78,15 @@ public function getClientAttribute()
...
+                ->where('status', 'unpaid');
```
Comment (bohraji) [inline app/Models/Finance/Invoice.php]: This will also be a query scope

File: app/Models/Finance/Invoice.php [hunk 6]
```diff
@@ -78,12 +78,15 @@ public function getClientAttribute()
         return optional($this->projectStageBillings()->first()->projectStage->project)->client;
     }
 
-    public static function filterByEndDates($end, $paginated = false)
+    public static function filterByEndDate($date, $paginated = false)
```
Comment (bohraji) [inline app/Models/Finance/Invoice.php]: This function not only filters by the end date, but also filters by the unpaid invoices. We can have the filter for paid, unpaid as an optional parameter for the function like we currently have `paginated` var.

File: app/Models/Finance/Invoice.php (hunk 4 above)
Comment (Satendra-SR) [reply app/Models/Finance/Invoice.php]: yes, you are right. I should do it as it can be used further in the future.

**Pull request closed without merging**